src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+to objects of the underlying type, called the *representation type*.
+You can choose to use some members, omit others,
+replace others, and add new functionality, 
+all while getting compile time errors before runtime errors.
+
+The following example wraps the `int` type to create an extension type
+that only allows operations that make sense for ID numbers:
+
+```dart
+extension type IdNumber(int i) {
+  // Wraps the 'int' type's '<' operator:
+  operator <(IdNumber other) => i < other.i;
+  // Doesn't declare the '+' operator, for example,
+  // because addition does not make sense for ID numbers.
+}
+
+void main() {
+  // Without the discipline of an extension type,
+  // 'int' exposes ID numbers to unsafe operations:
+  int myUnsafeId = 42424242;
+  myUnsafeId = myUnsafeId + 10; // OK, but shouldn't be allowed for IDs.
+
+  var safeId = IdNumber(42424242);
+  safeId + 10; // Compile-time error, no '+' operator.
+  myUnsafeId = safeId; // Compile-time error, wrong type.
+  myUnsafeId = safeId as int; // OK, runtime cast to representation type.
+}
+```
+
+{{site.alert.note}}
+Extension types serve the same purpose as **wrapper classes**,
+but don't require the creation of expensive runtime objects.
+Because extension types are static-only and compiled away at runtime,
+they are essentially zero cost.
+
+[**Extension methods**]() are a static abstraction similar to extension types.
+However, an extension method adds functionality to *all* instances
+of its underlying type. Extension types are the opposite; 
+an extension type's interface only applies to objects
+declared with the extension type,
+and is distinct from the interface of its underlying type by default. 
+Read the [Transparency](#transparency) section for more details.
+{{site.alert.end}}
+
+## Syntax
+
+### Declaration
+
+Define a new extension type with the `extension type` declaration and a name,
+followed by the *representation type declaration* in parenthesis:
+
+```dart
+extension type E(int i) {
+  // Define set of operations
+}
+```
+
+The representation type declaration specifies that 
+the underlying type of extension type `E` is `int`,
+and the reference to the *representation object* is named `i`.
+
+The representation object gives access to the underlying type,
+the same way declaring a final instance variable of the underlying type
+would in a wrapper class.
+It's in scope in the extension type body, and
+you can access it using its name as a getter:
+
+- Within the extension type body: `this.i`.
+- Outside with a property extraction: `e.i` (where `e` is the object of an extension type). 
+
+Extension types declarations can also include [type parameters](generics)
+just like classes:
+
+```dart
+extension type E<T>(int i) {
+  // ...
+}
+```
+
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
```

Comment:
Is it something to do with the `Name._` syntax that allows you to name the primary constructor, or is that just an arbitrary detail you and @lrhn happened to include in your named primary constructor examples? (I don't know what the `._` does there)

Reply:
Every constructor has a name. In a class/enum/extension-type/... declaration named `D`, a constructor can have the name `D`. Other constructors must then have a different name (in order to avoid name clashes), and we can use anything of the form `'D' '.' <identifier>` to do that, for example, `D.myName`.

With that in mind, some people like to say that the constructor named `D` is "unnamed", "nameless", "anonymous", or something like that, and they say that constructors with a name of the form `D.myName` are "named".

I prefer to just talk about the constructor and its name, never referring to any of them being "named" or "unnamed".

So the primary constructor has the name `E._` when it's declared as in `extension type E._(...) ...`, and it has the name `E` when it is declared as in `extension type E(...) ...`.

The reason why both Lasse and I chose `_` is probably that this is the shortest possible private name, and the whole point of using `extension type E._(...) ...` is that we _don't_ want to provide a constructor which is unconditionally creating an expression of the extension type based on an expression of the representation type.

In other words, we don't want to leave the door completely open for an introduction of the extension type, we want to have more control. For example:

```dart
int _currentIdNumber = 0;

extension type IdNumber._(int _value) {
  IdNumber(): this(_currentIdNumber++);
}
```

The intended and well-behaved approach is to call the public constructor with the name `IdNumber`, and that one does not accept any arguments because it is intended to simply use the next natural number that hasn't been used so far. It is still possible to do `42 as IdNumber`, but we should all know that this is a bad idea, and the fact that we can make the primary constructor private helps communicating this rule.

Reply:
Thank you very much for the detailed explanation!